build-scripts/multidirectory_test.sh

```diff
@@ -0,0 +1,229 @@
+#!/bin/sh
+set -eu
+
+create_test_repos() {
+    git init --quiet --bare origin.git
+    git init --quiet init/
+    cd init/
+    git commit --quiet --allow-empty --message first
+    git push --quiet ../origin.git master
+    cd -> /dev/null
+    git clone --quiet origin.git/ setup/
+}
+
+branches() {
+    (
+        cd setup/
+        master_rev=$(git rev-parse --short master)
+        for x; do
+            git checkout --quiet --orphan "$x"
+            git reset
+            git clean --quiet -df
+            > Jenkinsfile
+            git add Jenkinsfile
+            git commit --quiet \
+                --message "Auto-generated from $master_rev:jobs/$x"
+        done
+        git checkout --quiet master
+    )
+}
+
+jobs() {
+    (
+        cd setup/
+        for x; do mkdir -p "jobs/$x"; > "jobs/$x/Jenkinsfile"; done
+        git add .
+        git commit --quiet --message "$x: jobs"
+    )
+}
+
+expected_branches() {
+    (IFS=$(printf '\n ') && echo "$*" > expected_branches.txt)
+}
+
+execute_test() {
+    local branches x
+    (cd setup/ && git push --quiet --all)
+    git clone --quiet origin.git/ work/
+    (cd work/ && "$MULTIDIRECTORY_SH")
+    git clone --quiet origin.git/ check/
+    cd check/
+    branches=$(git ls-remote --heads origin \
+        | awk '{print gensub("^refs/heads/", "", 1, $2)}')
+    echo "$branches" | diff -u ../expected_branches.txt -
+    branches=$(echo "$branches" | { grep -x / || true; })
+    for x in $branches; do
+        ls "jobs/$x/Jenkinsfile" > /dev/null
+        (find * ! \( -wholename jobs -a -prune \) && cd "jobs/$x" && find *) \
```

Comment:
This is again to get the right output.  Since these are tests where all files
are under our control, I didn't think it was worth complicating this (already
non-trivial) line.